go/vm/lfvm/instructions_test.go

```diff
@@ -465,3 +466,116 @@ func TestMCopy(t *testing.T) {
 		})
 	}
 }
+
+func TestCreateShanghaiInitCodeSize(t *testing.T) {
+	maxInitCodeSize := uint64(49152)
+	tests := []struct {
+		revision       vm.Revision
+		init_code_size uint64
+		expected       Status
+	}{
+		{vm.R11_Paris, 0, RUNNING},
+		{vm.R11_Paris, 1, RUNNING},
+		{vm.R11_Paris, 2000, RUNNING},
+		{vm.R11_Paris, maxInitCodeSize - 1, RUNNING},
+		{vm.R11_Paris, maxInitCodeSize, RUNNING},
+		{vm.R11_Paris, maxInitCodeSize + 1, RUNNING},
+		{vm.R11_Paris, 100000, RUNNING},
+		{vm.R11_Paris, math.MaxUint64, RUNNING},
+
+		{vm.R12_Shanghai, 0, RUNNING},
+		{vm.R12_Shanghai, 1, RUNNING},
+		{vm.R12_Shanghai, 2000, RUNNING},
+		{vm.R12_Shanghai, maxInitCodeSize - 1, RUNNING},
+		{vm.R12_Shanghai, maxInitCodeSize, RUNNING},
+		{vm.R12_Shanghai, maxInitCodeSize + 1, MAX_INIT_CODE_SIZE_EXCEEDED},
+		{vm.R12_Shanghai, 100000, MAX_INIT_CODE_SIZE_EXCEEDED},
+		{vm.R12_Shanghai, math.MaxUint64, MAX_INIT_CODE_SIZE_EXCEEDED},
+	}
+
+	for _, test := range tests {
+		ctrl := gomock.NewController(t)
+		runContext := vm.NewMockRunContext(ctrl)
+
+		source := vm.Address{1}
+		ctxt := context{
+			status: RUNNING,
+			params: vm.Parameters{
+				Recipient: source,
+			},
+			context:  runContext,
+			stack:    NewStack(),
+			memory:   NewMemory(),
+			gas:      50000,
+			revision: test.revision,
+		}
+
+		// Prepare stack arguments.
+		ctxt.stack.stack_ptr = 3
+		ctxt.stack.data[0].Set(uint256.NewInt(test.init_code_size))
+
+		if test.expected == RUNNING {
+			runContext.EXPECT().Call(vm.Create, gomock.Any()).Return(vm.CallResult{}, nil)
+		}
+
+		opCreate(&ctxt)
+
+		if want, got := test.expected, ctxt.status; want != got {
+			t.Errorf("unexpected status after call, wanted %v, got %v", want, got)
+		}
+	}
+}
+
+func TestCreateShanghaiDeploymentCost(t *testing.T) {
+	static_gas := int64(32000)
+	tests := []struct {
+		revision       vm.Revision
+		init_code_size uint64
+		dynamic_cost   int64
+	}{
+		// gas cost from evm.codes
+		{vm.R11_Paris, 0, 32000 - static_gas},
+		{vm.R11_Paris, 1, 32003 - static_gas},
+		{vm.R11_Paris, 2000, 32196 - static_gas},
+		{vm.R11_Paris, 49152, 41216 - static_gas},
+
+		{vm.R12_Shanghai, 0, 32000 - static_gas},
+		{vm.R12_Shanghai, 1, 32005 - static_gas},
+		{vm.R12_Shanghai, 2000, 32322 - static_gas},
+		{vm.R12_Shanghai, 49152, 44288 - static_gas},
```

Comment:
nit: instead of hard-coded gas prices it would be nice to have a function for those to see how those values are derived.